File: 19 remove-nth-node-from-end-of-list/c/main.v2.c
/**
 * Leetcode 上的最佳解法
 */
struct ListNode *removeNthFromEnd(struct ListNode *head, int n) {
  struct ListNode *p = head, *pre = head, *pre1 = head;

  while (1) {
    while (n) {
      p = p->next;
      n--;
    }

    if (p == NULL) {
      if (pre == head) {
        head = pre->next;
      } else {
        pre1->next = pre->next;
      }
      return head;
    }
    pre1 = pre;
    pre = pre->next;
    p = p->next;
  }
}

File: 328 odd-even-linked-list/js/main.v2.js
/**
 *
 * Leetcode 最优解
 */

var oddEvenList = function(head) {
  if(head === null)
		return head;

  let h1 = null, // odd head
    p1 = null, // odd node
    p = head,
    t = null;

	while(p.next && p.next.next) {
    t = p.next.next;
		if(h1 === null) {
			h1 = p.next;
			p1 = h1;
		} else {
			p1.next = p.next;
			p1 = p1.next;
    }

		p.next = p.next.next;
		p = p.next;
  }

	if(p.next) {
		if(h1 === null) {
			h1 = p.next;
			p1 = h1;
		} else {
			p1.next = p.next;
			p1 = p1.next;
		}
  }

  if(p1) {
    p1.next = null;
  }
  p.next = h1;0

	return head;
};
